flink-ml-core/src/main/java/org/apache/flink/ml/common/sharedstorage/SharedStorageContextImpl.java

```diff
@@ -0,0 +1,117 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.common.sharedstorage;
+
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.StreamingRuntimeContext;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.util.function.BiConsumerWithException;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+/** Default implementation of {@link SharedStorageContext} using {@link SharedStorage}. */
+@SuppressWarnings("rawtypes")
+class SharedStorageContextImpl implements SharedStorageContext, Serializable {
+    private final StorageID storageID;
+    private final Map<ItemDescriptor, SharedStorage.Writer> writers = new HashMap<>();
+    private final Map<ItemDescriptor, SharedStorage.Reader> readers = new HashMap<>();
+    private Map<ItemDescriptor<?>, String> ownerMap;
+
+    public SharedStorageContextImpl() {
+        this.storageID = new StorageID();
+    }
+
+    public void setOwnerMap(Map<ItemDescriptor<?>, String> ownerMap) {
+        this.ownerMap = ownerMap;
+    }
+
+    @Override
+    public void invoke(BiConsumerWithException<SharedItemGetter, SharedItemSetter, Exception> func)
+            throws Exception {
+        func.accept(this::getSharedItem, this::setSharedItem);
+    }
+
+    private <T> T getSharedItem(ItemDescriptor<T> key) {
+        //noinspection unchecked
+        SharedStorage.Reader<T> reader = readers.get(key);
+        Preconditions.checkState(
+                null != reader,
+                String.format(
+                        "The operator requested to read a shared item %s not owned by itself.",
+                        key));
+        return reader.get();
+    }
+
+    private <T> void setSharedItem(ItemDescriptor<T> key, T value) {
+        //noinspection unchecked
+        SharedStorage.Writer<T> writer = writers.get(key);
+        Preconditions.checkState(
+                null != writer,
+                String.format(
+                        "The operator requested to read a shared item %s not owned by itself.",
+                        key));
+        writer.set(value);
+    }
+
+    @Override
+    public <T extends AbstractStreamOperator<?> & SharedStorageStreamOperator> void initializeState(
+            T operator,
+            StreamingRuntimeContext runtimeContext,
+            StateInitializationContext context) {
+        String ownerId = operator.getSharedStorageAccessorID();
+        int subtaskId = runtimeContext.getIndexOfThisSubtask();
+        for (Map.Entry<ItemDescriptor<?>, String> entry : ownerMap.entrySet()) {
+            ItemDescriptor descriptor = entry.getKey();
+            if (ownerId.equals(entry.getValue())) {
+                writers.put(
```

Comment:
There is no instances to be overwritten here, because every subtask owns a distinct instance of `SharedStorageContextImpl`.

The field of `context` in operators are serialized and then deserialized to different instances when operators being deployed. It just like other fields.